### ISSUE TITLE
[Rust] TVMError: Check failed: type_code_ == kDLInt (1 vs. 0) : expected int but get uint

### ISSUE THREAD

**Issue by 12101111** [apache/incubator-tvm#6197]
I got this when trying example of rust frontend:
```
original image dimensions: (256, 256)
resized image dimensions: (224, 224)
input size is [1, 3, 224, 224]
thread 'main' panicked at 'TVMError: Check failed: type_code_ == kDLInt (1 vs. 0) : expected int but get uint
Stack trace:
  File "../include/tvm/runtime/packed_func.h", line 383
  [bt] (0) /home/han/tvm/build/libtvm.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x6b) [0x7f577664915b]
  [bt] (1) /home/han/tvm/build/libtvm.so(tvm::runtime::TVMPODValue_::operator int() const+0x13e) [0x7f577664b58e]
  [bt] (2) /home/han/tvm/build/libtvm.so(tvm::runtime::GetAllContext(tvm::runtime::TVMArgs const&)+0xcb) [0x7f57772331bb]
  [bt] (3) /home/han/tvm/build/libtvm.so(+0x1118cfa) [0x7f5777234cfa]
  [bt] (4) /home/han/tvm/build/libtvm.so(TVMFuncCall+0x56) [0x7f57771d5196]
', tvm-rt/src/function.rs:123:9
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/libunwind.rs:86
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.46/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:78
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1117
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1508
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:198
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:217
  10: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:526
  11: rust_begin_unwind
             at src/libstd/panicking.rs:437
  12: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:391
  13: tvm_rt::function::Function::invoke
             at tvm-rt/src/function.rs:123
  14: resnet::main
             at tvm/examples/resnet/src/main.rs:75
  15: std::rt::lang_start::{{closure}}
             at /home/han/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  16: std::rt::lang_start_internal::{{closure}}
             at src/libstd/rt.rs:52
  17: std::panicking::try::do_call
             at src/libstd/panicking.rs:348
  18: std::panicking::try
             at src/libstd/panicking.rs:325
  19: std::panic::catch_unwind
             at src/libstd/panic.rs:394
  20: std::rt::lang_start_internal
             at src/libstd/rt.rs:51
  21: std::rt::lang_start
             at /home/han/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67
  22: main
  23: __libc_start_main
  24: _start
```

**Comment by 12101111** [comment]
This example code is incorrect.
```rust
    // get the global TVM graph runtime function
    let runtime_create_fn = Function::get("tvm.graph_runtime.create").unwrap();
    let runtime_create_fn_ret = runtime_create_fn.invoke(vec![
        graph.into(),
        (&lib).into(),
        (&ctx.device_type).into(),
        (&ctx.device_id).into(),
    ]);
```
The `into()` here don't convert `Context` to `ffi::DLContext`, so type of `ctx.device_id` is still  `usize`.
To confirm this, run the following code.
```rust
    let runtime_create_fn_ret = runtime_create_fn.invoke(vec![
        graph.into(),
        (&lib).into(),
        dbg!((&ctx.device_type).into()),
        dbg!((&ctx.device_id).into()),
    ]);
```
The output is:
```
[src/main.rs:37] (&ctx.device_type).into() = Int(
    1,
)
[src/main.rs:38] (&ctx.device_id).into() = UInt(
    0,
)
thread 'main' panicked at 'TVMError: Check failed: type_code_ == kDLInt (1 vs. 0) : expected int but get uint
```
This C++ type check macro expect type `int` but get `uint`.

And type of `ffi::DLContext` is wrong too. 
rust-bindgen generated this code:
```rust
#[doc = " \\brief The device type in DLContext."]
pub type DLDeviceType = u32;
#[doc = " \\brief A Device context for Tensor and operator."]
#[repr(C)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub struct DLContext {
    #[doc = " \\brief The device type used in the device."]
    pub device_type: DLDeviceType,
    #[doc = " \\brief The device index"]
    pub device_id: ::std::os::raw::c_int,
}
```
the type of `device_type`, `DLDeviceType` is `u32` in rust, but C++ side expect it's a `int`:
```rust
    let dlctx: ffi::DLContext = ctx.into();
    let runtime_create_fn_ret = runtime_create_fn.invoke(vec![
        graph.into(),
        (&lib).into(),
        dbg!((&dlctx.device_type).into()),
        dbg!((&dlctx.device_id).into()),
    ]);
```
```
[src/main.rs:37] (&dlctx.device_type).into() = UInt(
    1,
)
[src/main.rs:38] (&dlctx.device_id).into() = Int(
    0,
)
thread 'main' panicked at 'TVMError: Check failed: type_code_ == kDLInt (1 vs. 0) : expected int but get uint
```
The following code can pass the type check (and run into another error, which I should open a new issue for it)
```rust
let runtime_create_fn_ret = runtime_create_fn.invoke(vec![
        graph.into(),
        (&lib).into(),
        dbg!(1i32.into()),
        dbg!(0i32.into()),
    ]);
```
```
[src/main.rs:37] 1i32.into() = Int(
    1,
)
[src/main.rs:38] 0i32.into() = Int(
    0,
)
terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc
```
(This std::bad_alloc is cause by `load_param_fn.invoke(vec![barr.into()]).unwrap();`)

**Comment by jroesch** [comment]
@12101111 agreed I think this is another leftover from the previous version which I've been working on cleaning up and refactoring to bring up to date with modern TVM. My feeling is the graph runtime should use `tvm_sys::Context` as its public interface and only convert to the low level `tvm_sys::ffi` structs in a single location where we can be sure to correctly match the types as its very error prone and easy to pick the wrong integer types as they will work "most of the time".

**Comment by tqchen** [comment]
Fixed by  #6207
